### PR TITLE
[WPB-10092] open telemetry instrumentation 

### PR DESCRIPTION
**Pull request by MangoIV**
## TODO

- [x] wip: brig 
  - [x] incoming connections via wai
  - [x] outgoing connections intra 
- [x] wip: galley 
  - [x] incoming connections via wai
  - [x] outgoing connections intra 
- [x] gundeck
  - [x] incoming
  - [x] outgoing (cannon)
- [x] cannon 

## problems 

- check if spans done by hand work - for some reasons the names don't show up in jaeger, I don't know what and why is going on there. 

## to see what's going on:

- start jaeger like this: 
  ```bash 
  docker run --rm --name jaeger \
        -e COLLECTOR_ZIPKIN_HOST_PORT=:9411 \
        -p 6831:6831/udp \
        -p 6832:6832/udp \
        -p 5778:5778 \
        -p 16686:16686 \
        -p 4317:4317 \
        -p 4318:4318 \
        -p 14250:14250 \
        -p 14268:14268 \
        -p 14269:14269 \
        -p 9411:9411 \
        jaegertracing/all-in-one:latest
  ```

- start your services/ integration tests 
- open http://localhost:16686/
- voila 

## some research

this that I have found about otel so I don't forget them 

- https://opentelemetry.io/docs/concepts/context-propagation/
- https://opentelemetry.io/docs/concepts/signals/traces/?ref=unzip
- https://www.redhat.com/en/blog/openshift-commons-briefing-82-distributed-tracing-with-jaeger-prometheus-on-kubernetes
- https://hackage.haskell.org/package/hs-opentelemetry-instrumentation-wai-0.1.0.0/docs/src/OpenTelemetry.Instrumentation.Wai.html#local-6989586621679045744
- https://hackage.haskell.org/package/hs-opentelemetry-sdk-0.0.3.6
- https://hackage.haskell.org/package/hs-opentelemetry-sdk-0.0.3.6/docs/OpenTelemetry-Trace.html#t:NewLink
- https://hackage.haskell.org/package/hs-opentelemetry-sdk-0.0.3.6/docs/OpenTelemetry-Trace.html#t:Tracer

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

**Comment by MangoIV** [comment]
I'm not entirely sure the http/2 stuff works already, need to investigate a bit more

File: services/brig/src/Brig/Run.hs [hunk 1]
```diff
@@ -80,15 +83,16 @@ import Wire.UserStore
...
     Async.async $
-      runBrigToIO e $
-        wrapHttpClient $
-          Queue.listen (e ^. internalEvents) $
-            liftIO . runBrigToIO e . liftSem . Internal.onEvent
+      inSpan tracer "brig-event-listener" defaultSpanArguments $
+        runBrigToIO e $
+          wrapHttpClient $
+            Queue.listen (e ^. internalEvents) $
+              liftIO . inSpan tracer "brig-internal-event" defaultSpanArguments {- TODO(mangoiv): needs more information here -} . runBrigToIO e . liftSem . Internal.onEvent
```
Comment (MangoIV) [inline services/brig/src/Brig/Run.hs:95]: I'm not sure whether we need this at all.

File: tools/stern/src/Stern/Intra.hs [hunk 2]
```diff
@@ -996,12 +996,13 @@ getOAuthClient :: OAuthClientId -> Handler OAuthClient
...
-      ( method GET
-          . Bilge.paths ["i", "oauth", "clients", toByteString' cid]
-      )
+    lift $
+      rpc'
+        "brig"
+        b
+        ( method GET
+            . Bilge.paths ["i", "oauth", "clients", toByteString' cid]
+        )
```
Comment (MangoIV) [inline tools/stern/src/Stern/Intra.hs:1005]: this changes semantics but I don't think it's problematic as this is a direct call to `rpc'` which is what we do normally outside of `ExceptT` as well.

File: services/brig/src/Brig/Run.hs [hunk 3]
```diff
@@ -15,11 +15,7 @@
 -- You should have received a copy of the GNU Affero General Public License along
 -- with this program. If not, see <https://www.gnu.org/licenses/>.
 
-module Brig.Run
-  ( run,
-    mkApp,
-  )
-where
+module Brig.Run (run, mkApp) where
```
Comment (elland) [inline services/brig/src/Brig/Run.hs:18]: Feels like unnecessary change for less readability/

```suggestion
module Brig.Run (run, mkApp) where
  ( run,
    mkApp,
  )
where
```

Comment (elland) [inline services/brig/src/Brig/Run.hs:18]: Feels like unnecessary change for less readability.

```suggestion
module Brig.Run (run, mkApp) where
  ( run,
    mkApp,
  )
where
```

File: services/cannon/src/Cannon/Run.hs [hunk 4]
```diff
@@ -15,11 +15,7 @@
 -- You should have received a copy of the GNU Affero General Public License along
 -- with this program. If not, see <https://www.gnu.org/licenses/>.
 
-module Cannon.Run
-  ( run,
-    CombinedAPI,
-  )
-where
+module Cannon.Run (run, CombinedAPI) where
```
Comment (elland) [inline services/cannon/src/Cannon/Run.hs:18]:
```suggestion
module Cannon.Run (run, CombinedAPI) where
  ( run,
    CombinedAPI,
  )
where
```

File: services/galley/src/Galley/Intra/Util.hs [hunk 5]
```diff
@@ -15,15 +15,11 @@
 -- You should have received a copy of the GNU Affero General Public License along
 -- with this program. If not, see <https://www.gnu.org/licenses/>.
 
-module Galley.Intra.Util
-  ( IntraComponent (..),
-    call,
-  )
-where
+module Galley.Intra.Util (IntraComponent (..), call) where
```
Comment (elland) [inline services/galley/src/Galley/Intra/Util.hs:18]:
```suggestion
module Galley.Intra.Util (IntraComponent (..), call) where
  ( IntraComponent (..),
    call,
  )
where
```

File: changelog.d/3-bug-fixes/remove-spam-from-nginx [hunk 6]
```diff
@@ -0,0 +1 @@
+removed spam from nginx (nginz) by using the new style http/2 directive
```
Comment (fisx) [inline changelog.d/3-bug-fixes/remove-spam-from-nginx:1]: "spam"? maybe "spourious log messages"?

File: changelog.d/2-features/open-telemetry-instrumentation-brig-galley-gundeck-cannon [hunk 7]
```diff
@@ -0,0 +1 @@
+added open telemetry instrumentation for brig, galley, gundeck and cannon
```
Comment (fisx) [inline changelog.d/2-features/open-telemetry-instrumentation-brig-galley-gundeck-cannon:1]: maybe explain the consequences? where is the data going? how can i disable it going anywhere?

Comment (fisx) [reply changelog.d/2-features/open-telemetry-instrumentation-brig-galley-gundeck-cannon:1]: this info should also go into the markdown docs.

**Review by fisx — APPROVED** [review]
looks good! let's see what ci says.

File: services/galley/src/Galley/Run.hs [hunk 8]
```diff
@@ -82,25 +85,28 @@ run opts = lowerCodensity $ do
 
   void $ Codensity $ Async.withAsync $ runApp env deleteLoop
   void $ Codensity $ Async.withAsync $ runApp env refreshMetrics
-  lift $ finally (runSettingsWithShutdown settings' app Nothing) (closeApp env)
+  lift $ inSpan tracer "galley" defaultSpanArguments {kind = Otel.Server} (runSettingsWithShutdown settings' app Nothing) `finally` closeApp env
```
[the remark this answers is not in the document]
Comment (MangoIV) [reply services/galley/src/Galley/Run.hs:88]: Yeah but that’s true for everything interesting that is happening outside the actual http server - we have to add more instrumentation gradually, I guess?

File: docs/src/developer/developer/open-telemetry.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+     -p 14268:14268 \
+     -p 14269:14269 \
+     -p 9411:9411 \
+     jaegertracing/all-in-one:latest
+   ```
+
+2. Start your services or run integration tests.
+3. Open the Jaeger UI at [http://localhost:16686/](http://localhost:16686/)
+
+## Relevant Resources
```
Comment (fisx) [inline docs/src/developer/developer/open-telemetry.md:42]: @MangoIV merged, but do you want to add a section here on how to use this for debugging as a developer?

Comment (MangoIV) [reply docs/src/developer/developer/open-telemetry.md:42]: Theoretically yes, practically it will have to wait for next week :)
